### PR TITLE
Newdav maintenancemode

### LINKED ISSUE #5738: [10.0.0] The ownCloud desktop client is temporarily removing most of my files locally when server gets into maintenance mode
### Expected behaviour
When switching the server into maintenance mode, the ownCloud desktop client should report it is not connected to the server and synchronization should stop/fail.

### Actual behaviour
When switching the server into maintenance mode, the ownCloud desktop client still reports it is connected to the server and deletes most of my local files.

I am using "local" external storage and I am synchronizing about 1,3G, if that does matter.

### Steps to reproduce
1. In the ownCloud desktop client, select "force Sync now"(?) ("Synchronisation jetzt erzwingen"). The desktop client will re-scan synchronized folders locally, which will take a while.

2. During the previous step, put the ownCloud server into maintenance mode, i. e. edit /var/www/owncloud/config/config.php and set maintenance to true.

3. The ownCloud desktop client will delete most local files.

### Server configuration
Operating system:
Ubuntu 16.10

Web server: 
Apache/2.4.18 (Ubuntu)

Database:
psql (PostgreSQL) 9.5.6

PHP version:
PHP 7.0.15-0ubuntu0.16.10.4

ownCloud version:
ownCloud	10.0.0 (stable)

Storage backend (external storage):
"Local" external storage

### Client configuration
Client version:
2.3.1 (build 6824)

Operating system:
Windows10

OS language:
german

Installation path of client:
C:\Program Files (x86)\ownCloud

### PR DESCRIPTION
**Pull request by PVince81**
## Description
Enable maintenance mode plugin also for the new DAV endpoint.

## Related Issue
Fixes https://github.com/owncloud/client/issues/5738

## How Has This Been Tested?
Integration tests.
Local test with a simple PROPFIND with maintenance mode enabled.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Tagged as critical because it causes data loss with "Local" external storage and potentially other system-wide storages.

Needs backport to stable9.1.

@SergioBertolinSG I need your help with behat: the `@AfterScenario` that disables maintenance mode after the test runs too late. So while the tests pass, the other `@AfterScenario` code fails because maintenance mode is still active at this point. Any ideas how to change the priority ?

@DeepDiver1975 please review the fix

### CODE REVIEW

**Review by DeepDiver1975 — APPROVED** [review]
:face_with_head_bandage:

**Comment by ckamm** [comment]
Works for me

File: tests/integration/config/behat.yml [hunk 1]
```diff
@@ -13,15 +13,15 @@ default:
               - admin
             regular_user_password: 123456
             mailhog_url: http://127.0.0.1:8025/api/v2/messages
+        - CommandLineContext:
+            baseUrl: http://localhost:8080
+            ocPath: ../../
```
Comment (PVince81) [inline tests/integration/config/behat.yml]: clever!

**Comment by PVince81** [comment]
Rebased for CI

**Comment by PVince81** [comment]
```
14:37:46     /var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-new-endpoint.feature:5
14:37:46     /var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-new-endpoint.feature:547
14:37:46     /var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-old-endpoint.feature:5
14:37:46     /var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-old-endpoint.feature:490
```

**Comment by PVince81** [comment]
one more remaining, will have a look:
```
/var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-old-endpoint.feature:490
```

**Comment by PVince81** [comment]
boohoo the tests from this file pass locally

**Comment by PVince81** [comment]
more exact output from Jenkins (which I do not see locally):
```
13:15:57   Scenario: Maintenance mode                        # /var/lib/jenkins/workspace/owncloud-core_core_PR-27821-CG35ITJCHSNDATF4YR5HHRU2J7EA77TPHL4EG53CZCPDPGIQ2GDA/tests/integration/features/webdav-related-old-endpoint.feature:490
13:15:58     Given maintenance mode is enabled               # CommandLineContext::maintenanceModeIs()
13:15:58 [Tue May  9 13:15:58 2017] ::1:54497 [207]: /remote.php/webdav
13:15:58     When Connecting to dav endpoint as user "admin" # FeatureContext::connectingToDavEndpointAsUser()
13:15:58     Then the HTTP status code should be "503"       # FeatureContext::theHTTPStatusCodeShouldBe()
13:15:58       Failed asserting that 207 matches expected '503'.
13:15:58 [Tue May  9 13:15:58 2017] ::1:54502 [207]: /remote.php/dav/systemtags/
13:15:58 [Tue May  9 13:15:58 2017] Login failed: 'user0' (Remote IP: '::1')
13:15:59 [Tue May  9 13:15:58 2017] ::1:54505 [401]: /remote.php/webdav/myFileToTag.txt
13:15:59 [Tue May  9 13:15:59 2017] ::1:54506 [404]: /remote.php/dav/addressbooks/users/admin/MyAddressbook
13:15:59 [Tue May  9 13:15:59 2017] ::1:54509 [404]: /remote.php/dav/calendars/admin/MyCalendar
```

**Comment by SergioBertolinSG** [comment]
It is connecting using "admin" but you have changed it here https://github.com/owncloud/core/pull/27821/commits/4787fb7e0165dcda32b4c2035c523fe6bd2cdad4#diff-dc84a6418f59a28b93942dd9dd005b6bR842

**Comment by PVince81** [comment]
@SergioBertolinSG but then it should fail locally ? Are you able to reproduce it ?

**Comment by PVince81** [comment]
> It is connecting using "admin" but you have changed it here owncloud/core@4787fb7#diff-dc84a6418f59a28b93942dd9dd005b6bR842

Connecting as "admin" in the maintenance tests is the correct scenario.

Tried re-checking out the branch locally, removing all extra apps, the test still doesn't fail.

**Comment by SergioBertolinSG** [comment]
Nevermind my previus comment.

It is working fine for me:

```
[Tue May  9 13:24:13 2017] Exception: {"Message":"HTTP\/1.1 503 System in maintenance mode.","Exception":"Sabre\\DAV\\Exception\\ServiceUnavailable","Code":0,"Trace":"#0 [internal function]: OCA\\DAV\\Connector\\Sabre\\MaintenancePlugin->checkMaintenanceMode(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#1 \/var\/www\/html\/trashbin_vincent\/lib\/composer\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\n#2 \/var\/www\/html\/trashbin_vincent\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(466): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#3 \/var\/www\/html\/trashbin_vincent\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#4 \/var\/www\/html\/trashbin_vincent\/apps\/dav\/appinfo\/v1\/webdav.php(63): Sabre\\DAV\\Server->exec()\n#5 \/var\/www\/html\/trashbin_vincent\/remote.php(165): require_once('\/var\/www\/html\/t...')\n#6 {main}","File":"\/var\/www\/html\/trashbin_vincent\/apps\/dav\/lib\/Connector\/Sabre\/MaintenancePlugin.php","Line":83,"User":false}
[Tue May  9 13:24:13 2017] 127.0.0.1:33919 [503]: /remote.php/webdav
    When Connecting to dav endpoint as user "admin" # FeatureContext::connectingToDavEndpointAsUser()
```

**Comment by SergioBertolinSG** [comment]
I can reproduce this moving this scenario https://github.com/owncloud/core/pull/27821/files#diff-0d2db90e35aaf54e2c763f470420c026R547

into webdav-related-old-endpoint.feature.

I think the problem happens when running all the tests together. When running isolated it doesn't fail.

**Comment by PVince81** [comment]
So it's likely a side effect ? Maybe in some case the maintenance mode isn't properly disabled

**Comment by SergioBertolinSG** [comment]
I'm trying something, I'll tell you if it succeds.

**Comment by PVince81** [comment]
Tests are delayed now. We're experiencing env issues and potential race conditions.

So far it seems that while maintenance mode is enabled, the call still goes through when running the test on some env, and only sometimes. 1 of 5 test runs on the same env will fail... Suspecting PHP 7 bug with `proc_close` so far which maybe doesn't wait for the command to finish.

Needs further research...

**Comment by PVince81** [comment]
Observed so far:
- Jenkins has PHP 7.0.10 and has this failure
- @SergioBertolinSG's env has PHP 7.0.11 and it fails there too, but randomly
- my env has PHP 7.0.15 and I don't see any failure (had the tests running in a loop)

**Comment by SergioBertolinSG** [comment]
After adding this:

diff --git a/apps/dav/lib/Connector/Sabre/MaintenancePlugin.php b/apps/dav/lib/Connector/Sabre/MaintenancePlugin.php
index e648dce5e6..bf309b0aa0 100644
--- a/apps/dav/lib/Connector/Sabre/MaintenancePlugin.php
+++ b/apps/dav/lib/Connector/Sabre/MaintenancePlugin.php
@@ -76,6 +76,7 @@ class MaintenancePlugin extends ServerPlugin {
         * @return bool
         */
        public function checkMaintenanceMode() {
+       \OCP\Util::writeLog('DEBUG', 'maintenance: ' . $this->config->getSystemValue('maintenance', false), \OCP\Util::DEBUG);
                if ($this->config->getSystemValue('singleuser', false)) {
                        throw new ServiceUnavailable('System in single user mode.');
                }

And run with log level 0, I can see

```
And maintenance mode is enabled # CommandLineContext::maintenanceModeIs()
[Wed May 10 16:24:24 2017] maintenance: 
[Wed May 10 16:24:24 2017] 127.0.0.1:51562 [207]: /remote.php/dav/files/admin
When Connecting to dav endpoint as user "admin" # FeatureContext::connectingToDavEndpointAsUser()
Then the HTTP status code should be "503" # FeatureContext::theHTTPStatusCodeShouldBe()
Failed asserting that 207 matches expected '503'.
```

The enabling maintenance mode doesn't seems to work fine.

**Comment by PVince81** [comment]
Let's not waste more time on this testing issue and get the main fix merged.
We can continue researching this weird behavior in a separate PR that adds maintenance mode tests. I don't see these tests as really critical or helpful in the short term.

**Comment by PVince81** [comment]
I'll take care of the splitting tomorrow

**Comment by PVince81** [comment]
I've removed the test related commits from this PR, so only the fix remains.

The tests were moved to another branch and PR here https://github.com/owncloud/core/pull/27856

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/27857

**Comment by PVince81** [comment]
> [ERROR] Waited 600 seconds, no response

http://img.memecdn.com/we-waited-no-service_o_1937531.jpg

**Comment by PVince81** [comment]
rebased...

**Pull request merged**